### DISCUSSION TITLE
Proposal: Optimize delivery/encoding of query data response from /api/ds/query

### DISCUSSION THREAD

**Discussion by marefr** [grafana/grafana#44737] [Plugins]
**Background:**
When handling the [QueryData](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L22) method of a backend plugin the plugin developers are required to create and return a [QueryDataResponse](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L69-L72). The main result of that object is data frames ([data.Frames](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/data/frame.go#L70)) per query identifier/ref. [QueryDataResponse](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L69-L72) is the public SDK, but under the covers there's a [physical wire protocol](https://grafana.com/docs/grafana/latest/developers/plugins/backend/plugin-protocol/) which is built on gRPC and is defined in Protocol Buffers (a.k.a., protobuf). 

Currently, when Grafana calls the `QueryData` method of an backend plugin over gRPC, the data frames of the [QueryDataResponse](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L69-L72) is converted to Apache Arrow byte structure and sent back over the wire as bytes. When Grafana receives the gPRC response, it converts the Apache Arrow byte structure back to data frames and again creates a [QueryDataResponse](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L69-L72). This response is then used to stream Grafana HTTP API response as JSON ([code](https://github.com/grafana/grafana/blob/1f4fdf31dfc0eb2abfecc1e3817f372c2e08134e/pkg/api/metrics.go#L90-L99)).

For core/builtin data source plugins the behavior is a little bit different because communication happens in-memory and not over gPRC which means that conversion to Apache Arrow byte structure not happens. Core/builtin data source plugin creates [QueryDataResponse](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L69-L72). This response is then used to stream Grafana HTTP API response as JSON ([code](https://github.com/grafana/grafana/blob/1f4fdf31dfc0eb2abfecc1e3817f372c2e08134e/pkg/api/metrics.go#L90-L99)).

For any background service in Grafana in need of querying datasources, e.g. Grafana Alerting and Server Side Expressions, they operates on [data.Frames](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/data/frame.go#L70). 

**Problem:**
There's two different use cases regarding [QueryDataResponse](https://github.com/grafana/grafana-plugin-sdk-go/blob/a62a98d0ea262e968f84970dc9fce2ae4969da34/backend/data.go#L69-L72) and depending on which we would like to limit number of conversions happening back and forth of the response before serving it to the one that initiated it. 

_Update:_ Main point is that there's room for optimization in regards to providing an optimal result, based on who the consumer is, and by that limit the amount of conversions happening back and forth between data frames and Apache Arrow byte structure. This should in theory decrease the amount of memory needed to handle a data source query.  

**Proposal 1 - Use JSON bytes rather than Apache Arrow byte structure:**
In the case of initiating a query against Grafana's HTTP API (`/api/ds/query`), either from Grafana UI or either via curl or similar tools, the conversion between Apache Arrow byte structure and data frames is not optimal from a memory consumption perspective, see https://github.com/grafana/grafana/issues/43369 as an example. In this proposal we suggests converting the raw bytes sent from plugin to Grafana as JSON bytes and those bytes could be written directly to Grafana's `/api/ds/query` response (as JSON). 

Above based on some discussions with @ryantxu

**Comment by mckn** [comment]
This is awesome! Thanks for posting this 🙌🏻 

> There's two different use cases regarding QueryDataResponse and depending on which we would like to limit number of conversions happening back and forth of the response before serving it to the one that initiated it.

Can you please elaborate a bit more on why this is a problem?

**Reply by marefr** [reply]
Oops sorry I'll add that. But it's somewhat related to the proposal with unnecessary conversions and memory usage increase happening

**Reply by marefr** [reply]
Added an _Update_ paragraph.

**Comment by gabor** [comment]
hi, cool ideas here..

i'm a little getting lost in the transformations described, i will try to write them down again, please tell me if i got it wrong:

1. not-core data sources currently, dashboard query:
[QueryData returns QueryDataResponse] -> [Arrow] -> [QueryDataResponse] -> [JSON streamed to the browser]

2. core data sources currently, dashboard query:
[QueryData returns QueryDataResponse]  -> [JSON streamed to the browser]


regarding proposal1:

1. i'm not sure if the memory-problems in https://github.com/grafana/grafana/issues/43369 are caused by `Arrow->Dataframes` ... we don't 100% know for sure why the memory problem exists, but a big part is the hmm.. `native_prometheus_json_output->Dataframes` conversion, which is necessary, but currently uses a lot of extra memory. @ryantxu has a PoC solution to that specific problem at https://github.com/grafana/grafana/pull/44520
2. if i understand correctly, the proposal would do this: [QueryData returns json_bytes] -> [json_bytes streamed to the browser] .. what is the "json_bytes" format? bytes encoded as a base64 string in json?

**Reply by gabor** [reply]
( i updated the part where i talk about the `native_prometheus_json_output->Dataframes` conversion to provide more details)

**Reply by marefr** [reply]
1. Good to know :+1: 
2. data frames converted to JSON and the bytes of this

**Reply by ryantxu** [reply]
@gabor 

> what is the "json_bytes" format? bytes encoded as a base64 string in json?

With this proposal, the JSON bytes the client needs in the browser would be produced directly by the plugin, so that the plugin manager could avoid first decoding then re-encoding them.  With prometheus+loki embedded inside grafana, this makes no difference.  It only really matters for things that first travel over gRPC.

However as we discuss implementing prometheus+loki as "remote plugins" -- this change will make the grafana server resource overhead identical to what we had before where it is simply a `[]byte` proxy that never needs to decode and understand the results

**Comment by toddtreece** [comment]
@ryantxu and i discussed something like this last week in the context of caching, and it does seem like a potential improvement, but how would we address backward compatibility for external plugins? i'm assuming this would be a change to the SDK to convert to JSON instead of arrow, so it seems like it would be possible in that case for the plugin to send JSON to a version of grafana that's expecting arrow. would something equivalent to an `Accept` header work?

> For core/builtin data source plugins the behavior is a little bit different because communication happens in-memory and not over gPRC which means that conversion to Apache Arrow byte structure not happens

based on that info, it doesn't seem like this change would impact https://github.com/grafana/grafana/issues/43369 would it? 

there's also this somewhat related proposal for caching: https://github.com/grafana/grafana-enterprise/issues/2790

**Comment by kylebrandt** [comment]
+1 , if we can work out compatibility, I don't see a benefit to keeping arrow for our use cases - I think the conversion is purely technical debt at this point.

**Comment by ryantxu** [comment]
@marefr This sounds great!  When we built the JSON export, we considered this -- but punted because the core would decode then reincode anyway.

The only issue I see is that we want to be sure that plugins can still *send* native arrow if they must.  For rust/java/node/C# plugins, this is super easy.

The related -- but not yet useful -- option we should consider is having the encoding be part of the initial plugin load parameters.  Moving forward we should default to "JSON", but even in the future, if a rust ([datafusion](https://github.com/apache/arrow-datafusion)!) plugin wants to use plugins as a source, it will need to consume arrow

**Reply by marefr** [reply]
I started to explore this a while ago and decided to open a draft/exploratory PR of how this possibly could work https://github.com/grafana/grafana-plugin-sdk-go/pull/466

**Comment by ryantxu** [comment]
For decoding the bytes -- it may be enough to just inspect the first byte for `{` and then pick the right behavior.  Alternatively we could add an encoding option to the response, but that feels unnecessary

**Comment by toddtreece** [comment]
this might be a topic for a separate discussion, but i'm wondering if it would be worth adding protobufs (or something similar) for the data frame models, and generating structs from those? that way plugins written in other languages will have an easier time with generating the correct JSON structure

**Comment by mellieA** [comment]
This discussion has been closed in favor of https://github.com/grafana/grafana/issues/75147
